File: web/src/types/agent.ts
export type AgentStatus = 'active' | 'idle' | 'error' | 'paused';

export interface Agent {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  avatar_url?: string;
  avatar_emoji?: string;
  is_social_active?: boolean;
  status: AgentStatus;
  model: string;
  system_prompt: string;
  temperature: number;
  max_tokens: number;
  schedule_type: 'manual' | 'cron' | 'event';
  schedule_cron?: string;
  schedule_event?: string;
  scenario?: Record<string, any> | null;
  created_at: string;
  updated_at: string;
  tg_thread_id?: number | null;
  tg_group_id?: string | null;
  bot_username?: string | null;
  // UI helpers (computed client-side or from tasks)
  tasksToday?: number;
  avatarIcon?: string;
  avatarColor?: string;
}

export interface ModelInfo {
  id: string;
  label: string;
  provider: string;
  context_window: number;
  supports_tools: boolean;
  tier: 'free' | 'premium';
  description: string;
  available: boolean;
  badge?: string;
}

export interface RunResult {
  task_id: string;
  output: string;
  tools_used: string[];
  tokens_used: number;
  status: string;
  model: string;
  provider: string;
}

export interface Task {
  id: string;
  agent_id: string;
  status: 'queued' | 'running' | 'success' | 'failed' | 'cancelled';
  input_data?: Record<string, any>;
  output_data?: Record<string, any>;
  error_msg?: string;
  tokens_used: number;
  duration_ms?: number;
  created_at: string;
  updated_at: string;
  started_at?: string;
  completed_at?: string;
}

File: server/app/schemas/agent.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, field_validator, Field
from datetime import datetime

from app.core.models_registry import is_valid_model


class AgentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    description: Optional[str] = Field(None, max_length=512)
    avatar_url: Optional[str] = None
    avatar_emoji: Optional[str] = Field(None, max_length=16)
    is_social_active: bool = False
    model: str
    system_prompt: Optional[str] = Field(
        default="You are a helpful AI assistant.", max_length=100_000
    )
    temperature: Optional[float] = Field(default=0.7, ge=0.0, le=2.0)
    max_tokens: Optional[int] = Field(default=2048, ge=1, le=128_000)
    schedule_type: Optional[str] = "manual"
    schedule_cron: Optional[str] = None
    schedule_event: Optional[str] = None

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Agent name must not be blank")
        return v

    @field_validator("model")
    @classmethod
    def model_must_be_supported(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Model must not be blank")
        if not is_valid_model(v):
            from app.core.models_registry import SUPPORTED_MODELS
            raise ValueError(
                f"Model '{v}' is not supported. "
                f"Available: {sorted(SUPPORTED_MODELS.keys())}"
            )
        return v


class AgentCreate(AgentBase):
    pass


class AgentUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=64)
    description: Optional[str] = Field(None, max_length=512)
    avatar_url: Optional[str] = None
    avatar_emoji: Optional[str] = Field(None, max_length=16)
    is_social_active: Optional[bool] = None
    status: Optional[str] = None
    model: Optional[str] = None
    system_prompt: Optional[str] = Field(None, max_length=100_000)
    temperature: Optional[float] = Field(None, ge=0.0, le=2.0)
    max_tokens: Optional[int] = Field(None, ge=1, le=128_000)
    schedule_type: Optional[str] = None
    schedule_cron: Optional[str] = None
    schedule_event: Optional[str] = None

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("Agent name must not be blank")
        return v

    @field_validator("model")
    @classmethod
    def model_must_be_supported(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            if not is_valid_model(v):
                from app.core.models_registry import SUPPORTED_MODELS
                raise ValueError(
                    f"Model '{v}' is not supported. "
                    f"Available: {sorted(SUPPORTED_MODELS.keys())}"
                )
        return v


class Agent(AgentBase):
    id: str
    user_id: str
    status: str
    scenario: Optional[Dict[str, Any]] = None
    tg_thread_id: Optional[int] = None
    tg_group_id: Optional[str] = None
    bot_username: Optional[str] = None
    avatar_emoji: Optional[str] = None
    is_social_active: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Model info response schemas ────────────────────────────────────────────────

class ModelInfo(BaseModel):
    id: str
    label: str
    provider: str
    context_window: int
    supports_tools: bool
    tier: str
    description: str
    available: bool
    in_plan: bool
    status: Optional[str] = "active"
    latency: Optional[float] = 0.0
    last_checked: Optional[datetime] = None



class ModelsResponse(BaseModel):
    models: List[ModelInfo]
    user_has_custom_keys: bool
